src/messages/Lifecycle.ts

```diff
@@ -0,0 +1,141 @@
+import { guard, mixed, object, string } from 'decoders';
+import { LabeledMsg } from './LabeledMsg';
+import { createMessageValidator } from './validationUtils';
+
+/**
+ * Lifecycle stage in which the client
+ * notifies the host that they have started
+ * and ready to recieve messages.
+ */
+export interface LifecycleStarted {
+  stage: 'started';
+}
+
+/**
+ * Lifecycle stage in which the client
+ * notifies the host that they have stopped.
+ */
+export interface LifecycleStopped {
+  stage: 'stopped';
+}
+
+/**
+ * Environmental data provided to all clients
+ * in order to match behavior of the host application.
+ */
+export interface EnvData {
+  locale: string;
+}
+
+/**
+ * Lifecycle stage where environmental data
+ * is sent to the client.
+ */
+export interface LifecycleEnvironmentInit {
+  stage: 'env_init';
+  data: EnvData;
+}
+
+/**
+ * A lifecycle message for coordination between host and client.
+ *
+ * @remarks
+ * These are for internal use only.
+ */
+export type LifecycleStage =
+  | LifecycleStarted
+  | LifecycleStopped
+  | LifecycleEnvironmentInit;
```

Comment:
I'm wondering if we need an explicit life-cycle concept or one this complex. It feels to me like we could get away with just a simple request for environment context and a response from the host. That should be enough for the client to get itself up and running. What other purposes would these life-cycle stages serve?

Reply:
@justinaray had suggested a more complex lifecycle since his background workers used something like that.

I am ok with tuning this back a bit.  However, doing it this way gives us a bit more flexibility for the future.  More than a direct get environment, send environment message would be.

Reply:
I'm on board with structuring the messaging so we could grow to a more complex lifecycle if we need to. I'm just hesitant to build in the concept of a full lifecycle until we know we need it. I think that's one of those concepts that's open-ended enough that we run the risk of folks latching onto it to solve problems that would be better handled in other ways, leading to a lot of coupling between client and host state that will be difficult to maintain. 

For now I think its' good if we have clients send an event to indicate it's ready to start receiving input from the host, but I'd call it something like Ready or Initialized or Started and leave out the Lifecycle bit for now. That way in the future if we really need to, we can grow to something like:

```
export type LifecycleStage = 
    | Ready
    | Stopped
    | ...
```

but we're not building anything conceptually into the API that implies that growth.

Reply:
I think the latest changes should be more in line with what you were thinking.